rfd/0222-bot-instances-at-scale.md

```diff
@@ -0,0 +1,367 @@
+
+We have historically kept deprecated features around for longer than promised, to mitigate the risk of users missing these warnings, which creates a compounding maintenance burden.
+
+Instead, we will surface deprecation warnings and other such important messages in the web product by having `tbot` include them in its heartbeats to the auth server. It’s important to note that we will only send notices that **require some user action** or intervention, in order to avoid information fatigue — collecting the full stream of logs is out of scope (and would require a very different architecture).
+
+Services will receive a `NoticeLogger` along with their `readyz.Reporter`, it will “tag” all notices with the name and type of the service.
+
+```go
+// NoticeLogger allows a service to emit a notice that will be sent in the bot's
+// next heartbeat.
+type NoticeLogger interface {
+	// DeprecationWarning notifies the user that they're using a feature or
+	// configuration optional that will be removed in a future release.
+	//
+	// removalVersion identifies the major version in which the feature will
+	// be removed.
+	DeprecationWarning(message string, removalVersion string)
+
+	// Message sends a generic message.
+	Message(title string, body string)
+}
+```
+
+Notices emitted using the `NoticeLogger` will be buffered in-memory and flushed to the auth server on the bot’s next heartbeat. Once notices have been sent to the auth server, `tbot` will delete them from memory. To avoid unbounded memory growth if the heartbeat service becomes unavailable, notices will be stored in fixed-size data structure such as a ring buffer, although the volume of notices should be low enough for this not to be a problem anyway.
+
+On the auth server, notices will be stored on the bot instance record like all other heartbeat data. The server will keep a limited number (e.g. 50) of the most recent notices and discard the rest. We do not need to worry about concurrent writers because the `tbot`'s heartbeat service is a singleton and except in rare misconfigurations, it’s only possible for a single `tbot` process to use a bot instance’s identity at a time.
+
+If the heartbeat message’s `is_startup` flag is set, the auth server will discard all previous notices so that if the user fixes their configuration and restarts `tbot`, any warnings from their previous bad configuration will be immediately cleared to avoid confusion. This decision will also prevent us from treating notices as a general logging solution, as they will inherently be ephemeral and only representative of the most recent run.
+
+### Configuration
+
+For visibility, `tbot` will also send its full configuration to the auth server. As we do not yet support dynamically reloading `tbot`'s configuration (e.g. by sending a `SIGHUP`) we will only include it in the first heartbeat after the bot starts up, but this may change if we support dynamic configuration in the future.
```

Comment:
Will we store the configuration in a separate field on the BotInstance resource? Otherwise I think when the old heartbeats are overwritten we'll lose the configuration. We could use a similar behaviour as we do for notices.